File: java/src/main/java/com/agentlib/Agent.java
package com.agentlib;

import com.agentlib.skills.SkillInfo;
import com.agentlib.skills.SkillLoader;
import com.agentlib.skills.SkillTool;
import com.agentlib.tools.BashTool;
import com.agentlib.tools.CurlTool;
import com.agentlib.tools.GlobTool;
import com.agentlib.tools.GrepTool;
import com.agentlib.tools.ReadFileTool;
import com.agentlib.tools.SubAgentTool;
import com.agentlib.tools.UpdateFileTool;
import com.agentlib.tools.WriteFileTool;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.http.HttpClient;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Flow;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;

/**
 * Agent with lifecycle management (run/pause/resume/stop).
 * Emits events via Flow.Publisher for each step of execution.
 */
public class Agent {
    private final AgentConfig config;
    private final Map<String, Tool> tools;
    private final List<Message> messageHistory;
    private int turnCount;
    private long startTime;
    private final ObjectMapper objectMapper;
    private final ExecutorService executor;
    private SkillLoader skillLoader;

    // Lifecycle state
    private volatile AgentState state;
    private final Object stateLock = new Object();
    private final AtomicBoolean pauseFlag;
    private final AtomicBoolean stopFlag;

    // Callback
    private Consumer<Event> callback;

    /**
     * Agent lifecycle states.
     */
    public enum AgentState {
        IDLE, RUNNING, PAUSED, STOPPING, TERMINATED, COMPLETED
    }

    public Agent(AgentConfig config) {
        this.config = config;
        this.tools = new ConcurrentHashMap<>();
        this.messageHistory = Collections.synchronizedList(new ArrayList<>());
        this.turnCount = 0;
        this.objectMapper = new ObjectMapper();
        this.executor = Executors.newCachedThreadPool();
        this.state = AgentState.IDLE;
        this.pauseFlag = new AtomicBoolean(true); // not paused
        this.stopFlag = new AtomicBoolean(false);

        // Register built-in tools
        registerTool(new ReadFileTool());
        registerTool(new WriteFileTool());
        registerTool(new UpdateFileTool());
        registerTool(new GrepTool());
        registerTool(new GlobTool());
        registerTool(new BashTool(config.bashWhitelist(), config.bashBlacklist()));
        registerTool(new CurlTool(
            HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(30))
                .followRedirects(HttpClient.Redirect.NORMAL)
                .build(),
            config.curlWhitelist(),
            config.curlBlacklist()
        ));

        // Register custom tools
        for (Tool tool : config.customTools()) {
            registerTool(tool);
        }

        // Register subagent tool if enabled
        if (config.enableSubagent()) {
            registerTool(new SubAgentTool(config, messageHistory, tools));
        }

        // Initialize skills system if enabled
        if (config.enableSkills()) {
            this.skillLoader = new SkillLoader(
                config.skillsDir(),
                config.includeProjectSkills(),
                config.skillsProjectDir()
            );
            registerTool(new SkillTool(skillLoader));
        }
    }

    // ------------------------------------------------------------------
    // State management
    // ------------------------------------------------------------------

    public AgentState getState() {
        return state;
    }

    /**
     * Pause the agent. The current turn will finish, then pause.
     */
    public void pause() {
        synchronized (stateLock) {
            if (state == AgentState.RUNNING) {
                state = AgentState.PAUSED;
                pauseFlag.set(false);
            }
        }
    }

    /**
     * Resume a paused agent.
     */
    public void resume() {
        synchronized (stateLock) {
            if (state == AgentState.PAUSED) {
                state = AgentState.RUNNING;
                pauseFlag.set(true);
                stateLock.notifyAll();
            }
        }
    }

    /**
     * Stop the agent. Cannot be resumed.
     */
    public void stop() {
        synchronized (stateLock) {
            state = AgentState.STOPPING;
            stopFlag.set(true);
            pauseFlag.set(true); // unblock if paused
            stateLock.notifyAll();
        }
    }

    private boolean checkState() {
        AgentState current = state;
        return current != AgentState.STOPPING && current != AgentState.TERMINATED;
    }

    private void waitIfPaused() throws InterruptedException {
        synchronized (stateLock) {
            while (state == AgentState.PAUSED && !stopFlag.get()) {
                stateLock.wait(100);
            }
        }
    }

    // ------------------------------------------------------------------
    // Callback
    // ------------------------------------------------------------------

    public void setCallback(Consumer<Event> callback) {
        this.callback = callback;
    }

    private void emitEvent(Event event) {
        if (callback != null) {
            try {
                callback.accept(event);
            } catch (Exception e) {
                // Callback errors should not break the agent
            }
        }
    }

    // ------------------------------------------------------------------
    // Tool management
    // ------------------------------------------------------------------

    public void registerTool(Tool tool) {
        tools.put(tool.name(), tool);
    }

    public void unregisterTool(String name) {
        tools.remove(name);
    }

    public List<Tool> listTools() {
        return List.copyOf(tools.values());
    }

    // ------------------------------------------------------------------
    // History
    // ------------------------------------------------------------------

    public List<Message> getMessageHistory() {
        return messageHistory;
    }

    public void clearHistory() {
        messageHistory.clear();
        turnCount = 0;
    }

    // ------------------------------------------------------------------
    // Core methods
    // ------------------------------------------------------------------

    /**
     * Run the agent with the given input.
     * Returns a Flow.Publisher that emits events during execution.
     */
    public Flow.Publisher<Event> run(String input) {
        messageHistory.add(Message.user(input));
        startTime = System.currentTimeMillis();
        return runLoop();
    }

    /**
     * Continue the conversation with existing messages.
     */
    public Flow.Publisher<Event> chat(List<Message> messages) {
        messageHistory.addAll(messages);
        startTime = System.currentTimeMillis();
        return runLoop();
    }

    /**
     * Run the agent asynchronously and return the final content.
     */
    public CompletableFuture<String> runAsync(String input) {
        CompletableFuture<String> future = new CompletableFuture<>();
        Flow.Publisher<Event> publisher = run(input);

        publisher.subscribe(new Flow.Subscriber<>() {
            private Flow.Subscription subscription;
            private String finalContent = "";

            @Override
            public void onSubscribe(Flow.Subscription subscription) {
                this.subscription = subscription;
                subscription.request(Long.MAX_VALUE);
            }

            @Override
            public void onNext(Event event) {
                if (event instanceof CompleteEvent ce) {
                    finalContent = ce.finalContent();
                } else if (event instanceof ErrorEvent ee) {
                    future.completeExceptionally(new AgentException(ee.message()));
                }
            }

            @Override
            public void onError(Throwable throwable) {
                future.completeExceptionally(throwable);
            }

            @Override
            public void onComplete() {
                if (!future.isDone()) {
                    future.complete(finalContent);
                }
            }
        });

        return future;
    }

    // ------------------------------------------------------------------
    // Agent loop
    // ------------------------------------------------------------------

    private Flow.Publisher<Event> runLoop() {
        return subscriber -> {
            executor.submit(() -> {
                try {
                    synchronized (stateLock) {
                        state = AgentState.RUNNING;
                        startTime = System.currentTimeMillis();
                        stopFlag.set(false);
                        pauseFlag.set(true);
                    }

                    // Load skills for prompt
                    List<SkillInfo> skills = List.of();
                    if (skillLoader != null) {
                        skills = skillLoader.discoverAll();
                    }

                    String systemPrompt = PromptBuilder.build(
                        tools.values(),
                        config.systemPrompt(),
                        config.enableSubagent(),
                        config.subagentMaxTurns(),
                        skills
                    );

                    AnthropicClient client = new AnthropicClient(config);
                    int consecutiveCompactFailures = 0;

                    while (turnCount < config.maxTurns()) {
                        // Check stop
                        if (!checkState()) {
                            break;
                        }

                        // Wait if paused
                        waitIfPaused();
                        if (!checkState()) {
                            break;
                        }

                        turnCount++;
                        TurnStartEvent turnEvent = new TurnStartEvent(turnCount);
                        subscriber.onNext(turnEvent);
                        emitEvent(turnEvent);

                        // Check duration limit
                        if (config.maxDurationMs() > 0 && startTime > 0) {
                            long elapsed = System.currentTimeMillis() - startTime;
                            if (elapsed > config.maxDurationMs()) {
                                ErrorEvent error = new ErrorEvent("Max duration exceeded");
                                subscriber.onNext(error);
                                emitEvent(error);
                                break;
                            }
                        }

                        // Auto compact check
                        if (config.autoCompact()) {
                            AutoCompact.CompactResult result = AutoCompact.autoCompactIfNeeded(
                                client, config, List.copyOf(messageHistory), consecutiveCompactFailures
                            );
                            if (result.history().size() != messageHistory.size()) {
                                // Replace history with compacted version
                                messageHistory.clear();
                                messageHistory.addAll(result.history());
                                CompactEvent compactEvent = new CompactEvent(
                                    result.history().size(),
                                    AutoCompact.estimateTokens(result.history())
                                );
                                subscriber.onNext(compactEvent);
                                emitEvent(compactEvent);
                            }
                            consecutiveCompactFailures = result.failures();
                        }

                        // Build tool definitions
                        List<ToolDefinition> toolDefs = tools.values().stream()
                            .map(t -> new ToolDefinition(t.name(), t.description(), t.inputSchema()))
                            .toList();

                        // Wait if paused before API call
                        waitIfPaused();
                        if (!checkState()) {
                            break;
                        }

                        // Stream API response
                        List<ContentBlock> assistantContent = new ArrayList<>();
                        MessageStartEvent msgStart = new MessageStartEvent();
                        subscriber.onNext(msgStart);
                        emitEvent(msgStart);

                        try {
                            CountDownLatch streamLatch = new CountDownLatch(1);
                            List<StreamEvent> streamEvents = Collections.synchronizedList(new ArrayList<>());

                            Flow.Publisher<StreamEvent> streamPublisher = client.streamMessages(
                                List.copyOf(messageHistory),
                                systemPrompt,
                                toolDefs
                            );

                            streamPublisher.subscribe(new Flow.Subscriber<>() {
                                private Flow.Subscription subscription;

                                @Override
                                public void onSubscribe(Flow.Subscription subscription) {
                                    this.subscription = subscription;
                                    subscription.request(Long.MAX_VALUE);
                                }

                                @Override
                                public void onNext(StreamEvent event) {
                                    streamEvents.add(event);

                                    // Check stop during streaming
                                    if (!checkState()) {
                                        subscription.cancel();
                                        return;
                                    }

                                    try {
                                        waitIfPaused();
                                    } catch (InterruptedException e) {
                                        Thread.currentThread().interrupt();
                                        subscription.cancel();
                                        return;
                                    }

                                    String eventType = event.type();
                                    JsonNode data = event.data();

                                    if ("content_block_start".equals(eventType)) {
                                        ContentBlock block = parseBlock(data.get("content_block"));
                                        if (block != null) {
                                            assistantContent.add(block);
                                        }
                                    } else if ("content_block_delta".equals(eventType)) {
                                        JsonNode delta = data.get("delta");
                                        if (delta != null) {
                                            String deltaType = delta.get("type").asText();
                                            if ("text_delta".equals(deltaType)) {
                                                String text = delta.get("text").asText();
                                                MessageDeltaEvent mde = new MessageDeltaEvent(text);
                                                subscriber.onNext(mde);
                                                emitEvent(mde);
                                            } else if ("thinking_delta".equals(deltaType)) {
                                                String thinking = delta.get("thinking").asText();
                                                ThinkingDeltaEvent tde = new ThinkingDeltaEvent(thinking);
                                                subscriber.onNext(tde);
                                                emitEvent(tde);
                                            }
                                        }
                                    } else if ("message_stop".equals(eventType)) {
                                        // Message complete
                                    }
                                }

                                @Override
                                public void onError(Throwable throwable) {
                                    streamLatch.countDown();
                                }

                                @Override
                                public void onComplete() {
                                    streamLatch.countDown();
                                }
                            });

                            streamLatch.await(config.timeout().toMillis(), TimeUnit.MILLISECONDS);

                        } catch (Exception e) {
                            ErrorEvent error = new ErrorEvent("API error: " + e.getMessage());
                            subscriber.onNext(error);
                            emitEvent(error);
                            break;
                        }

                        MessageEndEvent msgEnd = new MessageEndEvent();
                        subscriber.onNext(msgEnd);
                        emitEvent(msgEnd);

                        if (!checkState()) {
                            break;
                        }

                        // Add assistant message to history
                        messageHistory.add(new Message(Role.ASSISTANT, List.copyOf(assistantContent)));

                        // Extract tool uses
                        List<ToolUseBlock> toolUses = assistantContent.stream()
                            .filter(b -> b instanceof ToolUseBlock)
                            .map(b -> (ToolUseBlock) b)
                            .toList();

                        if (toolUses.isEmpty()) {
                            // Task complete
                            String finalText = extractText(assistantContent);
                            synchronized (stateLock) {
                                state = AgentState.COMPLETED;
                            }
                            CompleteEvent complete = new CompleteEvent(finalText);
                            subscriber.onNext(complete);
                            emitEvent(complete);
                            return;
                        }

                        // Execute tools
                        List<ToolResultBlock> results = executeTools(toolUses, subscriber);

                        if (!checkState()) {
                            break;
                        }

                        // Add tool results to history
                        messageHistory.add(new Message(Role.USER,
                            results.stream().map(r -> (ContentBlock) r).toList()));
                    }

                    // Max turns reached
                    if (turnCount >= config.maxTurns()) {
                        ErrorEvent error = new ErrorEvent("Max turns reached");
                        subscriber.onNext(error);
                        emitEvent(error);
                    }

                } catch (Exception e) {
                    ErrorEvent error = new ErrorEvent(e.getMessage());
                    subscriber.onNext(error);
                    emitEvent(error);
                } finally {
                    synchronized (stateLock) {
                        if (state != AgentState.COMPLETED) {
                            state = AgentState.TERMINATED;
                        }
                    }
                    subscriber.onComplete();
                }
            });
        };
    }

    // ------------------------------------------------------------------
    // Tool execution
    // ------------------------------------------------------------------

    private List<ToolResultBlock> executeTools(
        List<ToolUseBlock> toolUses,
        Flow.Subscriber<? super Event> subscriber
    ) {
        // Group by read-only
        List<ToolUseBlock> readOnly = new ArrayList<>();
        List<ToolUseBlock> write = new ArrayList<>();

        for (ToolUseBlock tu : toolUses) {
            Tool tool = tools.get(tu.name());
            if (tool != null && tool.isReadOnly()) {
                readOnly.add(tu);
            } else {
                write.add(tu);
            }
        }

        List<ToolResultBlock> results = Collections.synchronizedList(new ArrayList<>());

        // Concurrent execution of read-only tools
        if (!readOnly.isEmpty()) {
            List<CompletableFuture<Void>> futures = readOnly.stream()
                .map(tu -> CompletableFuture.runAsync(() -> {
                    ToolResultBlock result = executeSingleTool(tu, subscriber);
                    results.add(result);
                }, executor))
                .toList();

            try {
                CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
                    .get(config.timeout().toMillis(), TimeUnit.MILLISECONDS);
            } catch (Exception e) {
                for (ToolUseBlock tu : readOnly) {
                    boolean found = results.stream().anyMatch(r -> r.toolUseId().equals(tu.id()));
                    if (!found) {
                        results.add(new ToolResultBlock(tu.id(),
                            "Tool execution failed: " + e.getMessage(), true));
                    }
                }
            }
        }

        // Sequential execution of write tools
        for (ToolUseBlock tu : write) {
            if (!checkState()) {
                break;
            }
            results.add(executeSingleTool(tu, subscriber));
        }

        return results;
    }

    private ToolResultBlock executeSingleTool(
        ToolUseBlock tu,
        Flow.Subscriber<? super Event> subscriber
    ) {
        ToolUseStartEvent startEvent = new ToolUseStartEvent(tu.name(), tu.id(), tu.input());
        subscriber.onNext(startEvent);
        emitEvent(startEvent);

        Tool tool = tools.get(tu.name());
        if (tool == null) {
            ToolResultBlock errorResult = new ToolResultBlock(tu.id(),
                "Tool not found: " + tu.name(), true);
            ToolUseEndEvent endEvent = new ToolUseEndEvent(tu.name(), tu.id(),
                new ToolResult(errorResult.content(), true));
            subscriber.onNext(endEvent);
            emitEvent(endEvent);
            return errorResult;
        }

        Map<String, Object> input = objectMapper.convertValue(tu.input(), new TypeReference<>() {});
        ToolContext ctx = new ToolContext(config.workDir(), List.copyOf(messageHistory),
            config.allowedReadDirs(), config.allowedWriteDirs());

        ToolResult result;
        try {
            result = tool.call(input, ctx);
        } catch (Exception e) {
            result = ToolResult.error(e.getMessage());
        }

        ToolUseEndEvent endEvent = new ToolUseEndEvent(tu.name(), tu.id(), result);
        subscriber.onNext(endEvent);
        emitEvent(endEvent);

        return new ToolResultBlock(tu.id(), result.content(), result.isError());
    }

    // ------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------

    private String extractText(List<ContentBlock> blocks) {
        StringBuilder sb = new StringBuilder();
        for (ContentBlock block : blocks) {
            if (block instanceof TextBlock tb) {
                sb.append(tb.text());
            }
        }
        return sb.toString();
    }

    private ContentBlock parseBlock(JsonNode data) {
        if (data == null) return null;
        String type = data.get("type").asText();
        return switch (type) {
            case "text" -> new TextBlock(data.get("text").asText());
            case "tool_use" -> new ToolUseBlock(
                data.get("name").asText(),
                data.get("id").asText(),
                data.get("input")
            );
            case "thinking" -> new ThinkingBlock(
                data.get("thinking").asText(),
                data.has("signature") ? data.get("signature").asText() : null
            );
            default -> null;
        };
    }
}

File: nodejs/src/agent.ts
/** Agent core with lifecycle management. */

import { AnthropicClient, ToolDefinition } from './client.js';
import {
  AgentConfig, ITool, resolveConfig, ToolContext,
} from './config.js';
import { autoCompactIfNeeded, estimateTokens } from './compact.js';
import { buildSystemPrompt } from './prompt.js';
import {
  ContentBlock, Event, isToolUse, Message,
  userMessage, assistantMessage, messageStartEvent, messageEndEvent,
  messageDeltaEvent, thinkingDeltaEvent, turnStartEvent, errorEvent, completeEvent,
  compactEvent,
} from './types.js';
import { ReadFileTool, WriteFileTool, UpdateFileTool, BashTool, CurlTool, GlobTool, GrepTool, SubAgentTool } from './tools/index.js';
import { SkillLoader, SkillTool } from './skills/index.js';
import type { SkillInfo } from './skills/types.js';

type ToolUseBlock = Extract<ContentBlock, { type: 'tool_use' }>;
type TextBlock = Extract<ContentBlock, { type: 'text' }>;

export enum AgentState {
  IDLE = 'idle',
  RUNNING = 'running',
  PAUSED = 'paused',
  STOPPING = 'stopping',
  TERMINATED = 'terminated',
  COMPLETED = 'completed',
}

export class Agent {
  private config: Required<AgentConfig>;
  private tools: Map<string, ITool> = new Map();
  private messageHistory: Message[] = [];
  private turnCount = 0;
  private startTime: number | null = null;
  private skillLoader: SkillLoader | null = null;
  private loadedSkills: SkillInfo[] = [];

  private _state = AgentState.IDLE;
  private pauseResolve: (() => void) | null = null;
  private pausePromise: Promise<void> | null = null;
  private stopped = false;

  constructor(config: AgentConfig) {
    this.config = resolveConfig(config);
    this.initTools();
  }

  private initTools() {
    this.registerTool(new ReadFileTool());
    this.registerTool(new WriteFileTool());
    this.registerTool(new UpdateFileTool());
    this.registerTool(new BashTool(this.config.bashWhitelist, this.config.bashBlacklist));
    this.registerTool(new CurlTool(this.config.curlWhitelist, this.config.curlBlacklist));
    this.registerTool(new GlobTool());
    this.registerTool(new GrepTool());
    for (const tool of this.config.tools) this.registerTool(tool);

    if (this.config.enableSubagent) {
      this.registerTool(new SubAgentTool(this.config, this.messageHistory, this.tools));
    }

    if (this.config.enableSkills) {
      this.skillLoader = new SkillLoader({
        skillsDir: this.config.skillsDir,
        includeProjectSkills: this.config.includeProjectSkills,
        projectDir: this.config.skillsProjectDir,
      });
      this.registerTool(new SkillTool(this.skillLoader));
    }
  }

  get state(): AgentState { return this._state; }

  pause(): void {
    if (this._state === AgentState.RUNNING) {
      this._state = AgentState.PAUSED;
      this.pausePromise = new Promise((resolve) => { this.pauseResolve = resolve; });
    }
  }

  resume(): void {
    if (this._state === AgentState.PAUSED) {
      this._state = AgentState.RUNNING;
      this.pauseResolve?.();
      this.pauseResolve = null;
      this.pausePromise = null;
    }
  }

  stop(): void {
    this._state = AgentState.STOPPING;
    this.stopped = true;
    this.pauseResolve?.();
    this.pauseResolve = null;
    this.pausePromise = null;
  }

  private checkState(): boolean {
    return this._state !== AgentState.STOPPING && this._state !== AgentState.TERMINATED;
  }

  private async waitIfPaused(): Promise<void> {
    if (this._state === AgentState.PAUSED && this.pausePromise) {
      await this.pausePromise;
    }
  }

  registerTool(tool: ITool): void { this.tools.set(tool.name, tool); }
  unregisterTool(name: string): void { this.tools.delete(name); }
  listTools(): ITool[] { return Array.from(this.tools.values()); }
  getMessageHistory(): Message[] { return [...this.messageHistory]; }
  clearHistory(): void { this.messageHistory = []; this.turnCount = 0; }

  async *run(input: string): AsyncGenerator<Event> {
    this.messageHistory.push(userMessage(input));
    yield* this.runLoop();
  }

  async *chat(messages: Message[]): AsyncGenerator<Event> {
    this.messageHistory.push(...messages);
    yield* this.runLoop();
  }

  private async *runLoop(): AsyncGenerator<Event> {
    this._state = AgentState.RUNNING;
    this.startTime = Date.now();
    this.stopped = false;

    // Load skills lazily on first run loop
    if (this.config.enableSkills && this.skillLoader && this.loadedSkills.length === 0) {
      this.loadedSkills = this.skillLoader.discoverAll();
    }

    const systemPrompt = buildSystemPrompt(
      this.tools,
      this.config.systemPrompt,
      this.config.enableSubagent,
      this.config.subagentMaxTurns,
      this.loadedSkills,
    );
    const client = new AnthropicClient(this.config);
    let consecutiveCompactFailures = 0;

    try {
      while (this.turnCount < this.config.maxTurns) {
        if (!this.checkState()) break;
        await this.waitIfPaused();
        if (!this.checkState()) break;

        this.turnCount++;
        yield turnStartEvent(this.turnCount);
        this.config.callback?.(turnStartEvent(this.turnCount));

        if (this.config.maxDurationMs > 0 && this.startTime) {
          const elapsed = Date.now() - this.startTime;
          if (elapsed > this.config.maxDurationMs) {
            yield errorEvent('Max duration exceeded');
            break;
          }
        }

        const toolDefs: ToolDefinition[] = Array.from(this.tools.values()).map(t => ({
          name: t.name,
          description: t.description,
          input_schema: t.inputSchema,
        }));

        // Auto compact
        if (this.config.autoCompact) {
          const result = await autoCompactIfNeeded(this.config, client, this.messageHistory, consecutiveCompactFailures);
          if (result.history.length !== this.messageHistory.length) {
            this.messageHistory = result.history;
            const estimated = estimateTokens(this.messageHistory);
            const e = compactEvent(this.messageHistory.length, estimated);
            yield e;
            this.config.callback?.(e);
          }
          consecutiveCompactFailures = result.failures;
        }

        await this.waitIfPaused();
        if (!this.checkState()) break;

        const assistantContent: ContentBlock[] = [];
        yield messageStartEvent();
        this.config.callback?.(messageStartEvent());

        try {
          for await (const event of client.streamMessages(this.messageHistory, systemPrompt, toolDefs)) {
            if (!this.checkState()) break;
            await this.waitIfPaused();

            const type = event.type as string;
            if (type === 'content_block_start') {
              assistantContent.push(this.parseBlock(event.content_block as Record<string, unknown>));
            } else if (type === 'content_block_delta') {
              const delta = event.delta as Record<string, unknown>;
              if (delta.type === 'text_delta') {
                const e = messageDeltaEvent(delta.text as string);
                yield e;
                this.config.callback?.(e);
              } else if (delta.type === 'thinking_delta') {
                const e = thinkingDeltaEvent(delta.thinking as string);
                yield e;
                this.config.callback?.(e);
              }
            } else if (type === 'message_stop') {
              yield messageEndEvent();
              this.config.callback?.(messageEndEvent());
            }
          }
        } catch (e) {
          yield errorEvent(`API error: ${e}`);
          break;
        }

        if (!this.checkState()) break;

        this.messageHistory.push(assistantMessage(assistantContent));
        const toolUses = assistantContent.filter(isToolUse);

        if (toolUses.length === 0) {
          const finalText = this.extractText(assistantContent);
          this._state = AgentState.COMPLETED;
          const e = completeEvent(finalText);
          yield e;
          this.config.callback?.(e);
          return;
        }

        const results = await this.executeTools(toolUses);
        if (!this.checkState()) break;
        this.messageHistory.push({ role: 'user', content: results });
      }

      if (this.turnCount >= this.config.maxTurns) {
        yield errorEvent('Max turns reached');
      }
    } finally {
      if (this._state !== AgentState.COMPLETED) {
        this._state = AgentState.TERMINATED;
      }
    }
  }

  private async executeTools(toolUses: ToolUseBlock[]): Promise<ContentBlock[]> {
    const readOnly: ToolUseBlock[] = [];
    const write: ToolUseBlock[] = [];
    for (const block of toolUses) {
      const tool = this.tools.get(block.name);
      if (tool?.isReadOnly) readOnly.push(block); else write.push(block);
    }

    const results: ContentBlock[] = [];

    if (readOnly.length > 0) {
      const readResults = await Promise.all(readOnly.map(b => this.executeSingleTool(b)));
      results.push(...readResults);
    }

    for (const block of write) {
      if (!this.checkState()) break;
      results.push(await this.executeSingleTool(block));
    }

    return results;
  }

  private async executeSingleTool(block: ToolUseBlock): Promise<ContentBlock> {
    const tool = this.tools.get(block.name);
    if (!tool) {
      return { type: 'tool_result', tool_use_id: block.id, content: `Tool not found: ${block.name}`, is_error: true };
    }

    const context: ToolContext = {
      workDir: this.config.workDir,
      messageHistory: [...this.messageHistory],
    };

    try {
      const result = await tool.call(block.input, context);
      return { type: 'tool_result', tool_use_id: block.id, content: result.content, is_error: result.isError };
    } catch (e) {
      return { type: 'tool_result', tool_use_id: block.id, content: String(e), is_error: true };
    }
  }

  private extractText(blocks: ContentBlock[]): string {
    return blocks.filter((b): b is TextBlock => b.type === 'text').map((b) => b.text).join('');
  }

  private parseBlock(data: Record<string, unknown>): ContentBlock {
    const type = data.type as string;
    if (type === 'text') return { type: 'text', text: data.text as string };
    if (type === 'tool_use') return { type: 'tool_use', name: data.name as string, id: data.id as string, input: (data.input as Record<string, unknown>) || {} };
    if (type === 'thinking') return { type: 'thinking', thinking: data.thinking as string };
    throw new Error(`Unknown block type: ${type}`);
  }
}
